### PR TITLE
Feat: Enable kotlin 2

### PR DESCRIPTION
**Pull request by nym3r0s**
## Overview

* Attempt to allow jqwik to use K2 compiler

### Details

* Remove JSR305 annotations
* Remove NonnullApi annotation -> replaced by `@NullMarked`

_List open issues and remaining problems with your solution._

---

I hereby agree to the terms of the [jqwik Contributor Agreement](https://github.com/jqwik-team/jqwik/blob/master/CONTRIBUTING.md#jqwik-contributor-agreement).

### CODE REVIEW

**Comment by jlink** [comment]
@nym3r0s Hi! I very much appreciate your effort to tackle the K2 compliance!

The problem I have with your pull request is that you also changed __a lot of__ formatting and order of methods etc in __many__ files. This makes it hard to impossible for me to go through all changes. I'm absolutely willing to consider a PR if it contains the minimum necessary changes only. Many thanks!

**Comment by nym3r0s** [comment]
@jlink sorry about that! looks like my editor didn't have the editorconfig present and re-formatted everything. I've updated it now - pls feel free to review. thanks!

File: api/build.gradle [hunk 1]
```diff
@@ -51,6 +51,7 @@ signing {
 }
 
 dependencies {
+	compileOnly("com.google.errorprone:error_prone_annotations:${errorproneVersion}")
```
Comment (jlink) [inline api/build.gradle:54]: What is this dependency needed for?

Comment (nym3r0s) [reply api/build.gradle:54]: Needed for `@CheckReturnValue`. As recommended [here](https://errorprone.info/bugpattern/CheckReturnValue#fnref:jsr)

**Comment by jlink** [comment]
Looks good as far as I can see now. I'll try it within the IDE an the latest Kotlin plugin and then get back to you - hopefully tomorrow.

**Review by jlink — APPROVED** [review]
Thanks a lot!
In case you also want to have a look at the remaining Kotlin compiler warnings, go ahead (in separate PR).

File: dependencies.gradle [hunk 2]
```diff
@@ -40,6 +40,7 @@ ext {
 	assertJVersion = '3.26.0'
 	mockitoVersion = '4.11.0'
 	systemLambdaVersion = '1.2.1'
+	errorproneVersion = '2.27.1'
 	kotlinVersion = '1.9.22' // TODO: Support version 2.x
```
[the remark this answers is not in the document]
Comment (nym3r0s) [reply dependencies.gradle:44]: The usage of `kotlinVersion` is commented out 😅 but I'll update this for hygiene. and yes, findbugsVersion is no longer needed.

[the remark this answers is not in the document]
Comment (nym3r0s) [reply dependencies.gradle]: Done! ✅

**Comment by vlsi** [comment]
By the way, when building Kotlin code, there's a possibility to target a specific Kotlin version.
For instance, one can configure `apiVersion=1.8`, then the code will be consumable by Kotlin 1.8+ clients.

I am not sure what is the default target when using Kotlin 2.0, however, I guess it would be better to target at least "latestRelease-1" or "latestRelease-2" to give users some time so they do not need to bump their Kotlin version every time they update jqwik.

---

Currently, there's `jvmTarget=...` (see https://github.com/jqwik-team/jqwik/blob/b8eb1c95a4653c0b9fd9bf778e995497f5aa21cb/kotlin/build.gradle#L80C16-L80C25), however, it would be better to use `-Xjdk-release=$jdkRelease` compiler option. Then Kotlin compiler would avoid issues like https://www.morling.dev/blog/bytebuffer-and-the-dreaded-nosuchmethoderror/

See https://github.com/apache/jmeter/blob/bacaa8baff2365a6b49d996766c77475c37eecc3/build-logic/jvm/src/main/kotlin/build-logic.kotlin.gradle.kts#L45-L55

The same goes for Java: it is better to stop using `sourceCompatibility/targetCompatibility` and use `-release` instead.
Then jqwik could use the latest compiler and still properly target the required Java version.

That is what we have in pgjdbc: https://github.com/pgjdbc/pgjdbc/blob/423ee03b6ec42b874101fbed2cb7f07973362c33/build-logic/jvm/src/main/kotlin/build-logic.java.gradle.kts#L57
We require a recent `javac` (e.g. 17+), and we target Java 8 thanks to `--release`.

**Comment by jlink** [comment]
> By the way, when building Kotlin code, there's a possibility to target a specific Kotlin version...

It will take me some time to figure out the most recent way to do that correctly in Gradle.

But I'm happy to accept a PR for that ;-)

**Comment by nym3r0s** [comment]
@jlink - could we pls release this version to maven repos? Thanks!

**Comment by jlink** [comment]
> @jlink - could we pls release this version to maven repos? Thanks!

It's available as snapshot version `1.9.0-SNAPSHOT`. For GA the "min target Kotlin version" issue should be resolved first.

**Pull request merged**